proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,112 @@
+
+C# 7.2 [introduced `in` parameters](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md#solution-in-parameters) as a way to pass readonly references.
+`in` parameters allow both lvalues and rvalues and can be used without any annotation at the callsite.
+However, APIs which capture or return references from their parameters would like to disallow rvalues and also enforce some indication at the callsite that a reference is being captured.
+`ref readonly` parameters are ideal in such cases as they warn if used with rvalues or without any annotation at the callsite.
+
+Furthermore, there are APIs that need only read-only references but use
+- `ref` parameters since they were introduced before `in` became available and changing to `in` would be a source and binary breaking change, e.g., `QueryInterface`, or
+- `in` parameters to accept readonly references even though passing rvalues to them doesn't really make sense, e.g., `ReadOnlySpan<T>..ctor(in T value)`, or
+- `ref` parameters to disallow rvalues even though they don't mutate the passed reference, e.g., `Unsafe.IsNullRef`.
+
+These APIs could migrate to `ref readonly` parameters without breaking users.
+For details on binary compatibility, see the proposed [metadata encoding](#metadata).
+Specifically, changing
+- `ref` → `ref readonly` would only be a binary breaking change for virtual methods,
+- `ref` → `in` would also be a binary breaking change for virtual methods, but not a source breaking change (because the rules change to only warn for `ref` arguments passed to `in` parameters),
+- `in` → `ref readonly` would not be a breaking change (but no callsite annotation or rvalue would result in a warning).
+
+In the opposite direction, changing
+- `ref readonly` → `ref` would be potentially a source breaking change (unless only `ref` callsite annotation was used), and a binary breaking change for virtual methods,
+- `ref readonly` → `in` would not be a breaking change (but `ref` callsite annotation would result in a warning).
+
+## Detailed design
+[design]: #detailed-design
+
+No changes in grammar are necessary.
+Specification will be extended to allow `ref readonly` modifiers for parameters with the same rules as specified for `in` parameters in [their proposal](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md), except where explicitly changed in this proposal.
+
+### Interchangeability with `in`
+
+Members declared in a single type cannot differ in signature solely by `ref`/`out`/`in`/`ref readonly`.
+For other purposes of signature matching (e.g., hiding or overriding), `ref readonly` can be interchanged with `in` modifier, but that results in a warning at the declaration site [[§7.6](https://github.com/dotnet/csharpstandard/blob/47912d4fdae2bb8c3750e6485bdc6509560ec6bf/standard/basic-concepts.md#76-signatures-and-overloading)].
```

Comment:
Similarly, here. In my opinion, for the feature to be really useful and to actually provide guarantees to API authors, the interchangeability between `in` and `ref readonly` in signatures should not be allowed.

Reply:
This remains a consideration for back-compat. We have 20 years of existing code taking `ref` and 3-5 years of code taking `in` that would prefer to switch to newer features but can't due to it being a breaking change.

`ref->in` is currently source breaking because `ref x` for an `in T` is invalid syntax. Allowing `ref x` to be used for `in T` avoids this break and fixes a weird quirk of an older compiler consuming metadata emitted by a newer compiler (https://github.com/dotnet/csharplang/issues/1308).

Likewise, the BCL has rejected using `in` for several APIs due to the hidden copies it allows and the high chance for accidental incorrect API usage in this scenario. A warning for such hidden copies with the `ref readonly` feature is considered sufficient to allow the transition and allows APIs (including APIs currently taking `in`) to transition without breaking users.